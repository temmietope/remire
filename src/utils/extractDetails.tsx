import { Result } from '../models/';
import { capitalize } from './capitalize';

const detailsWanted = [
  'name',
  'age',
  'color',
  'owned_by',
  'year',
  'hair_color',
  'vehicles',
  'birth_year',
  'eye_color',
  'films',
  'gender',
  'height',
  'homeworld',
  'mass',
  'skin_color',
  'created',
  'edited',
  'species',
  'starships',
  'vehicles',
  'title',
  'episode_id',
  'opening_crawl',
  'director',
  'producer',
  'release_date',
  'species',
  'planets',
  'model',
  'starship_class',
  'manufacturer',
  'cost_in_credits',
  'length',
  'crew',
  'passengers',
  'max_atmosphering_speed',
  'hyperdrive_rating',
  'MGLT',
  'cargo_capacity',
  'consumables',
  'pilots',
  'vehicle_class',
  'length',
  'films',
  'classification',
  'designation',
  'average_height',
  'average_lifespan',
  'eye_colors',
  'hair_colors',
  'skin_colors',
  'language',
  'homeworld',
  'surface_water',
  'residents',
  'orbital_period',
  'rotation_period',
];

type GenericObject = { [key: string]: any }

/**
 * @description extract required data from payload
 */
const extractDetails = (payload: GenericObject): Result[][] =>
  Object.entries(payload)
    .filter(([k]) => detailsWanted.includes(k))
    .map(([k, v]: any) => [capitalize(k), v]);

export default extractDetails;
