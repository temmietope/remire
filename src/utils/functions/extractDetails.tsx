import { convertToTitle } from "./convertToTitle";

const details_wanted = [
  "name",
  "age",
  "color",
  "owned_by",
  "year",
  "hair_color",
  "vehicles",
  "birth_year",
  "eye_color",
  "films",
  "gender",
  "height",
  "homeworld",
  "mass",
  "skin_color",
  "created",
  "edited",
  "species",
  "starships",
  "vehicles",
  "title",
  "episode_id",
  "opening_crawl",
  "director",
  "producer",
  "releases_date",
  "species",
  "planets",
  "model",
  "starship_class",
  "manufacturer",
  "cost_in_credits",
  "length",
  "crew",
  "passengers",
  "max_atmosphering_speed",
  "hyperdrive_rating",
  "MGLT",
  "cargo_capacity",
  "consumables",
  "pilots",
  "vehicle_class",
  "length",
  "films",
  "classification",
  "designation",
  "average_height",
  "average_lifespan",
  "eye_colors",
  "hair_colors",
  "skin_colors",
  "language",
  "homeworld",
  "surface_water",
  "residents",
  "orbital_period",
  "rotation_period",
];

// const template = {
//   name: "Name",
//   skin_color: "Skin color",
//   owned_by: "Owned By",
//   hair_color: "Hair color",
//   vehicles: "vehicles",
// };

// function titleCase(sentence:string):string {
//   let sentence = string.toLowerCase().split(" ");
//   for (let i = 0; i < sentence.length; i++) {
//     sentence[i] = sentence[i][0].toUpperCase() + sentence[i].slice(1);
//   }

//   return sentence.join(" ");
// }

type GenericObject = { [key: string]: any };

export function extractDetails(payload: GenericObject): GenericObject {
  return (
    Object.entries(payload)
      .filter(([k]) => details_wanted.includes(k))
      // .map(([k, v]: any) => ({ [template[k]]: v }));
      // .map(([k, v]: any) => ({ [convertToTitle(k)]: v }))
      .map(([k, v]: any) => [convertToTitle(k), v])
  );
}
