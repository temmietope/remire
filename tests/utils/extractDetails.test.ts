// @ts-nocheck
import extractDetails from '../../src/utils/extractDetails';

const list = [
  {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male',
    homeworld: 'http://swapi.dev/api/planets/1/',
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/2/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/6/',
    ],
    species: [],
    vehicles: [
      'http://swapi.dev/api/vehicles/14/',
      'http://swapi.dev/api/vehicles/30/',
    ],
    starships: [
      'http://swapi.dev/api/starships/12/',
      'http://swapi.dev/api/starships/22/',
    ],
    created: '2014-12-09T13:50:51.644000Z',
    edited: '2014-12-20T21:17:56.891000Z',
    url: 'http://swapi.dev/api/people/1/',
  },
  {
    name: 'CR90 corvette',
    model: 'CR90 corvette',
    manufacturer: 'Corellian Engineering Corporation',
    cost_in_credits: '3500000',
    length: '150',
    max_atmosphering_speed: '950',
    crew: '30-165',
    passengers: '600',
    cargo_capacity: '3000000',
    consumables: '1 year',
    hyperdrive_rating: '2.0',
    MGLT: '60',
    starship_class: 'corvette',
    pilots: [],
    films: [
      'http://swapi.dev/api/films/1/',
      'http://swapi.dev/api/films/3/',
      'http://swapi.dev/api/films/6/',
    ],
    created: '2014-12-10T14:20:33.369000Z',
    edited: '2014-12-20T21:23:49.867000Z',
    url: 'http://swapi.dev/api/starships/2/',
  },
  {
    name: 'T-16 skyhopper',
    model: 'T-16 skyhopper',
    manufacturer: 'Incom Corporation',
    cost_in_credits: '14500',
    length: '10.4 ',
    max_atmosphering_speed: '1200',
    crew: '1',
    passengers: '1',
    cargo_capacity: '50',
    consumables: '0',
    vehicle_class: 'repulsorcraft',
    pilots: [],
    films: ['http://swapi.dev/api/films/1/'],
    created: '2014-12-10T16:01:52.434000Z',
    edited: '2014-12-20T21:30:21.665000Z',
    url: 'http://swapi.dev/api/vehicles/6/',
  },
];

describe('should capitalize and replace underscore', () => {
  for (const o of list) {
    it('should transform object', () => {
      const output = extractDetails(o);

      expect(output).toMatchSnapshot();
    });
  }
});
