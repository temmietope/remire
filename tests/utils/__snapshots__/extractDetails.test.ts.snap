// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should capitalize and replace underscore should transform object 1`] = `
Array [
  Array [
    "Name",
    "Luke Skywalker",
  ],
  Array [
    "Height",
    "172",
  ],
  Array [
    "Mass",
    "77",
  ],
  Array [
    "Hair color",
    "blond",
  ],
  Array [
    "Skin color",
    "fair",
  ],
  Array [
    "Eye color",
    "blue",
  ],
  Array [
    "Birth year",
    "19BBY",
  ],
  Array [
    "Gender",
    "male",
  ],
  Array [
    "Homeworld",
    "http://swapi.dev/api/planets/1/",
  ],
  Array [
    "Films",
    Array [
      "http://swapi.dev/api/films/1/",
      "http://swapi.dev/api/films/2/",
      "http://swapi.dev/api/films/3/",
      "http://swapi.dev/api/films/6/",
    ],
  ],
  Array [
    "Species",
    Array [],
  ],
  Array [
    "Vehicles",
    Array [
      "http://swapi.dev/api/vehicles/14/",
      "http://swapi.dev/api/vehicles/30/",
    ],
  ],
  Array [
    "Starships",
    Array [
      "http://swapi.dev/api/starships/12/",
      "http://swapi.dev/api/starships/22/",
    ],
  ],
  Array [
    "Created",
    "2014-12-09T13:50:51.644000Z",
  ],
  Array [
    "Edited",
    "2014-12-20T21:17:56.891000Z",
  ],
]
`;

exports[`should capitalize and replace underscore should transform object 2`] = `
Array [
  Array [
    "Name",
    "CR90 corvette",
  ],
  Array [
    "Model",
    "CR90 corvette",
  ],
  Array [
    "Manufacturer",
    "Corellian Engineering Corporation",
  ],
  Array [
    "Cost in credits",
    "3500000",
  ],
  Array [
    "Length",
    "150",
  ],
  Array [
    "Max atmosphering speed",
    "950",
  ],
  Array [
    "Crew",
    "30-165",
  ],
  Array [
    "Passengers",
    "600",
  ],
  Array [
    "Cargo capacity",
    "3000000",
  ],
  Array [
    "Consumables",
    "1 year",
  ],
  Array [
    "Hyperdrive rating",
    "2.0",
  ],
  Array [
    "MGLT",
    "60",
  ],
  Array [
    "Starship class",
    "corvette",
  ],
  Array [
    "Pilots",
    Array [],
  ],
  Array [
    "Films",
    Array [
      "http://swapi.dev/api/films/1/",
      "http://swapi.dev/api/films/3/",
      "http://swapi.dev/api/films/6/",
    ],
  ],
  Array [
    "Created",
    "2014-12-10T14:20:33.369000Z",
  ],
  Array [
    "Edited",
    "2014-12-20T21:23:49.867000Z",
  ],
]
`;

exports[`should capitalize and replace underscore should transform object 3`] = `
Array [
  Array [
    "Name",
    "T-16 skyhopper",
  ],
  Array [
    "Model",
    "T-16 skyhopper",
  ],
  Array [
    "Manufacturer",
    "Incom Corporation",
  ],
  Array [
    "Cost in credits",
    "14500",
  ],
  Array [
    "Length",
    "10.4 ",
  ],
  Array [
    "Max atmosphering speed",
    "1200",
  ],
  Array [
    "Crew",
    "1",
  ],
  Array [
    "Passengers",
    "1",
  ],
  Array [
    "Cargo capacity",
    "50",
  ],
  Array [
    "Consumables",
    "0",
  ],
  Array [
    "Vehicle class",
    "repulsorcraft",
  ],
  Array [
    "Pilots",
    Array [],
  ],
  Array [
    "Films",
    Array [
      "http://swapi.dev/api/films/1/",
    ],
  ],
  Array [
    "Created",
    "2014-12-10T16:01:52.434000Z",
  ],
  Array [
    "Edited",
    "2014-12-20T21:30:21.665000Z",
  ],
]
`;
